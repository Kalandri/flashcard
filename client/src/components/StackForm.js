import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Form, FormGroup, FormControl, ControlLabel } from 'react-bootstrap';
import { addStack, saveStack } from '../actions';
import { Link } from 'react-router-dom';

export class StackForm extends Component {
  constructor() {
    super();

    this.state = { 
      title: '',
      cards: []
    }
  }


  addCard() {
    const { cards } = this.state;

    cards.push({ id: cards.length, prompt: '', answer: '' });

    this.setState({ cards });
  }

  addStack() {
    this.props.addStack(this.state);
    if(this.props.auth!==null && this.props.auth!==false){
     this.props.saveStack(this.state);
     console.log("state", this.state);
    }
  }

  updateCardPart = (event, index, part) => {
    const { cards } = this.state;
    
    cards[index][part] = event.target.value;

    this.setState({ cards });
  }

  render() {
    return (
      <div>
        <h4>Create a New Stack</h4>
        <br />
        <Form inline>
          <FormGroup>
            <ControlLabel>Title:</ControlLabel>
            {' '}
            <FormControl onChange={event => this.setState({ title: event.target.value })} />
          </FormGroup>
          {
            this.state.cards.map((card, index) => {
              return (
                <div key={card.id}>
                  <br />
                  <FormGroup>
                    <ControlLabel>Prompt:</ControlLabel>
                    {' '}
                    <FormControl 
                      onChange={event => this.updateCardPart(event, index, 'prompt')} 
                    />
                    {' '}
                    <ControlLabel>Answer:</ControlLabel>
                    {' '}
                    <FormControl 
                      onChange={event => this.updateCardPart(event, index, 'answer')}
                    />
                  </FormGroup>
                </div>
              )
            })
          }
        </Form>
        <div>
        <button onClick={() => this.addCard()} className="btn waves-effect waves-light btn-large  cyan darken-1">Add Card</button>
        </div>
        <div className="right">
        <Link to="/dashboard" className="btn waves-effect waves-light btn-large  cyan darken-3 left">Cancel</Link>
        <button onClick={()=>this.addStack()} className="btn waves-effect waves-light btn-large  cyan darken-3 right" type="submit" name="action"> Save and Add the Stack</button>
        </div>
      </div>
    )
  }
}

export default connect(null, { addStack, saveStack })(StackForm);