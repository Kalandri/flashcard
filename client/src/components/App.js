import React, { Component } from 'react';
import { connect } from 'react-redux';
import { BrowserRouter, Switch, Route } from 'react-router-dom';

import { fetchUser } from '../actions/index';
import Header from './Header';
import Dashboard from './Dashboard';
import Stack from './Stack';
import StackForm from './StackForm';
import Landing from './Landing';

import '../index.css';
import 'materialize-css/dist/css/materialize.min.css';
class App extends Component {
	componentDidMount(){
		this.props.fetchUser();
	}

	render(){
		return (
			<div >
			<BrowserRouter>
				<div>
					<Header />
				     <Switch>
				        <Route exact path='/' component={ Landing } />
				        <Route path='/dashboard' component = { Dashboard } />
				        <Route path='/stack' component={ Stack } />
				        <Route path='/stackform' component={ StackForm } />
				    </Switch>
		      	</div>
		    </BrowserRouter>
		    </div>
		);
	}
}

export default connect(null, { fetchUser })(App);