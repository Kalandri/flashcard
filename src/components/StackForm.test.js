import React, { Component } from 'react';
import App from './App';
import { MemoryRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import { StackForm } from  './StackForm';
import renderer from 'react-test-renderer';

const chTitle = 'change title';
const chPrompt ='change prompt';
const chAnswer = 'change answer';

describe('StackForm', ()=>{
	const stackForm = shallow(<StackForm />);

	const changeFormControl=(index, val)=>{
		stackForm.find('FormControl').at(index).simulate('change', {target: {value: val}});
	}

	describe('it render the structure as expected', ()=>{
		it('renders the form title', ()=>{
			expect(stackForm.find('h4').at(1).text()).toEqual('Create a New Stack');
		});

		it('renders a link home', ()=>{
			expect(stackForm.find('Link h4').text()).toEqual('Home');
		});

		it('renders a Form Component', ()=>{
			expect(stackForm.find('Form').exists()).toBe(true);
		});

		it('render a button to add a new card', ()=>{
			expect(stackForm.find('Button').at(0).props().children).toEqual('Add Card');
		});

		it ('render a button to submit the card', ()=>{
			expect(stackForm.find('Button').at(1).props().children).toEqual('Save and Add the Stack');
		});
	});
	
	describe('and updating the title', ()=>{
		beforeEach(()=>{
			changeFormControl(0, chTitle);
		});
		it('update the title in the state', ()=>{			
			expect(stackForm.state().title).toEqual(chTitle);
		});
	});

	describe('when adding a new card', ()=>{
		beforeEach(()=>{
			stackForm.find('Button').at(0).simulate('click');
		});

		afterEach(()=>{
			stackForm.setState({cards:[]});
		});

		it('add a new card to the state', ()=>{
			expect(stackForm.state().cards.length).toEqual(1);
		});

		it('renders the prompt section', ()=>{
			expect(stackForm.find('ControlLabel').at(1).props().children).toEqual('Prompt:');
		}); 

		it('renders the Answer section', ()=>{
			expect(stackForm.find('ControlLabel').at(2).props().children).toEqual('Answer:');
		});

		describe('and updating the card prompt', ()=>{
			beforeEach(()=>{ 
				changeFormControl(1, chPrompt); 
			});
			it('update the prompt in the state', ()=>{
				console.log(stackForm.state());
				expect(stackForm.state().cards[0].prompt).toEqual(chPrompt);
			});
		});

		describe('and updating the card answer', ()=>{
			 beforeEach(()=>{
			 	changeFormControl(2, chAnswer);
			 });

			 it('updae the answer in the state', ()=>{
			 	expect(stackForm.state().cards[0].answer).toEqual(chAnswer);
			 });
		});
	});
});

test('snapshot test for StackForm', ()=>{
	const component = renderer.create(
	    <MemoryRouter >
	      <StackForm />
	    </MemoryRouter>
  	);	
	const tree=component.toJSON();
	expect(tree).toMatchSnapshot();

});