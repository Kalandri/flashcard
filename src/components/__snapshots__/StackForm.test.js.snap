// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for StackForm 1`] = `
<div>
  <a
    className="link-home"
    href="/"
    onClick={[Function]}
  >
    <h4>
      Home
    </h4>
  </a>
  <h4>
    Create a New Stack
  </h4>
  <br />
  <form
    className="form-inline"
  >
    <div
      className="form-group"
    >
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Title:
      </label>
       
      <input
        className="form-control"
        id={undefined}
        onChange={[Function]}
        type={undefined}
      />
    </div>
  </form>
  <br />
  <button
    className="btn btn-default"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add Card
  </button>
   
  <button
    className="btn btn-default"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Save and Add the Stack
  </button>
</div>
`;
